protocol/localtxmonitor/client.go

```diff
@@ -0,0 +1,215 @@
+	acquiredSlot       uint64
+	acquireResultChan  chan bool
+	hasTxResultChan    chan bool
+	nextTxResultChan   chan []byte
+	getSizesResultChan chan MsgReplyGetSizesResult
+}
+
+// NewClient returns a new LocalTxMonitor client object
+func NewClient(protoOptions protocol.ProtocolOptions, cfg *Config) *Client {
+	if cfg == nil {
+		tmpCfg := NewConfig()
+		cfg = &tmpCfg
+	}
+	c := &Client{
+		config:             cfg,
+		acquireResultChan:  make(chan bool),
+		hasTxResultChan:    make(chan bool),
+		nextTxResultChan:   make(chan []byte),
+		getSizesResultChan: make(chan MsgReplyGetSizesResult),
+	}
+	// Update state map with timeout
+	stateMap := StateMap.Copy()
+	if entry, ok := stateMap[stateAcquiring]; ok {
+		entry.Timeout = c.config.AcquireTimeout
+		stateMap[stateAcquiring] = entry
+	}
+	if entry, ok := stateMap[stateBusy]; ok {
+		entry.Timeout = c.config.QueryTimeout
+		stateMap[stateBusy] = entry
+	}
+	// Configure underlying Protocol
+	protoConfig := protocol.ProtocolConfig{
+		Name:                protocolName,
+		ProtocolId:          protocolId,
+		Muxer:               protoOptions.Muxer,
+		ErrorChan:           protoOptions.ErrorChan,
+		Mode:                protoOptions.Mode,
+		Role:                protocol.ProtocolRoleClient,
+		MessageHandlerFunc:  c.messageHandler,
+		MessageFromCborFunc: NewMsgFromCbor,
+		StateMap:            stateMap,
+		InitialState:        stateIdle,
+	}
+	c.Protocol = protocol.New(protoConfig)
+	// Start goroutine to cleanup resources on protocol shutdown
+	go func() {
+		<-c.Protocol.DoneChan()
+		close(c.acquireResultChan)
+		close(c.hasTxResultChan)
+		close(c.nextTxResultChan)
+		close(c.getSizesResultChan)
+	}()
+	return c
+}
+
+func (c *Client) messageHandler(msg protocol.Message, isResponse bool) error {
+	var err error
+	switch msg.Type() {
+	case MessageTypeAcquired:
+		err = c.handleAcquired(msg)
+	case MessageTypeReplyHasTx:
+		err = c.handleReplyHasTx(msg)
+	case MessageTypeReplyNextTx:
+		err = c.handleReplyNextTx(msg)
+	case MessageTypeReplyGetSizes:
+		err = c.handleReplyGetSizes(msg)
+	default:
+		err = fmt.Errorf("%s: received unexpected message type %d", protocolName, msg.Type())
+	}
+	return err
+}
+
+func (c *Client) acquire() error {
+	msg := NewMsgAcquire()
+	if err := c.SendMessage(msg); err != nil {
+		return err
+	}
+	// Wait for reply
+	<-c.acquireResultChan
+	return nil
+}
+
+func (c *Client) release() error {
+	msg := NewMsgRelease()
+	if err := c.SendMessage(msg); err != nil {
+		return err
+	}
+	c.acquired = false
+	return nil
+}
+
+// Acquire starts the acquire process for a current mempool snapshot
+func (c *Client) Acquire() error {
+	c.busyMutex.Lock()
+	defer c.busyMutex.Unlock()
+	return c.acquire()
+}
+
+// Release releases the previously acquired mempool snapshot
+func (c *Client) Release() error {
+	c.busyMutex.Lock()
+	defer c.busyMutex.Unlock()
+	return c.release()
+}
+
+// Stop transations the protocol to the Done state. No more operations will be possible
```

Comment:
transitions

Reply:
This was also wrong in the file I copied it from, so I fixed it there too